tests/csapi/apidoc_room_state_test.go

```diff
@@ -104,26 +106,29 @@ func TestRoomState(t *testing.T) {
 				"preset":     "public_chat",
 			})
 
-			res := authedClient.MustDoFunc(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"})
-			foundRoom := false
-
-			must.MatchResponse(t, res, match.HTTPResponse{
-				JSON: []match.JSON{
-					match.JSONKeyPresent("chunk"),
-					match.JSONArrayEach("chunk", func(val gjson.Result) error {
-						gotRoomID := val.Get("room_id").Str
-						if gotRoomID == roomID {
-							foundRoom = true
+			authedClient.MustDoFuncUntil(t, "GET", []string{"_matrix", "client", "v3", "publicRooms"}, func(res *http.Response) bool {
+				foundRoom := false
+
+				must.MatchResponse(t, res, match.HTTPResponse{
```

Comment:
You can't use `MatchResponse` in this way as it will fail the test if the response does not pass the given matchers.

Reply:
(NB Diff is outdated but basic structure remains the same, so will reply to comment)

I believe this is what we want, though (or at least I'm used to it from SyTest); if the response is ever malformed then we want to fail the test immediately without waiting for it to be correct.

Put another way: The retries are waiting for the expected data to appear in the response, but that doesn't mean that any responses before that are allowed to omit required fields.
As a result, you end up with 3 outcomes: success, temporary fail (i.e. retry) and permanent fail.

As for how these currently seem to map to code:
Success = `return true`
Temporary fail = `return false`
Permanent fail = `t.FailNow`/`t.Fatalf`